gflownet/envs/base.py

```diff
@@ -295,6 +295,53 @@ def get_logprobs(
         logprobs = self.logsoftmax(logits)[ns_range, action_indices]
         return logprobs
 
+    def step_random(self):
+        """
+        Samples a random action and executes the step.
+
+        Returns
+        -------
+        state : list
+            The state after executing the action.
+
+        action : int
+            Action, randomly sampled.
+
+        valid : bool
+            False, if the action is not allowed for the current state.
+        """
+        mask_invalid = torch.unsqueeze(
+            torch.tensor(self.get_mask_invalid_actions_forward(), dtype=torch.bool), 0
+        )
+        random_policy = torch.unsqueeze(
+            torch.tensor(self.random_policy_output, dtype=self.float), 0
+        )
+        actions, _ = self.sample_actions(
+            policy_outputs=random_policy, mask_invalid_actions=mask_invalid
+        )
+        action = actions[0]
+        return self.step(action)
+
+    def trajectory_random(self):
+        """
+        Samples a random trajectory, by sampling random actions until an EOS action is
+        sampled.
+
+        Returns
+        -------
+        state : list
+            The final state.
+
+        action: list
+            The list of actions (tuples) in the trajectory.
+        """
+        actions = []
+        while self.done is not True:
+            _, action, valid = self.step_random()
+            if valid:
```

Comment:
I don't think so, because this depends on `step_random()`, which always obtains the mask and uses it to sample the random action. Do you agree?

Reply:
I agree with you which is why I raised that point. Therefore, I think that the `if` is unneeded and you can simply always append the action to the list.

Reply:
Ah I get you. This is a subtle point. I think that it's good to keep the notion of `valid` inside the `step()` method as separate from the notion of valid as per the masks. Why? For example, we know that computing the mask can be very expensive, because we need to compute validity for each action of possibly many. Therefore, in some environments the mask might not ultimately determine whether an action is valid, but rather it is designed to determine which actions are _invalid_, to filter out actions (hence the word "invalid" in the variable name). It is in the `step()` action, that validity is ultimately determined after _attempting_ to execute the action.

Long story short: I believe we should keep checking `valid` in `step()` and in `trajectory_random()` before adding to the list of actions.

Does that make sense? Do you agree?

Reply:
So you're saying that there are cases where an action might be invalid even though the mask says it is valid.

Is that only for environments with continuous actions spaces or there are other cases?

If it can happen on environments with discrete action spaces I wonder whether something should be changed in the code to handle those.
I'm just thinking of a scenario in which the GFN assigns almost all probability mass to an action which is valid according to the mask but that the environment refuses. It could be stuck in this loop of "get mask, sample action, attempt action unsucceccfully" for a while.

At any rate, even if this is an issue, I think that it falls outside of the scope of your current PR. I think that my original comment has been addressed.

Reply:
> So you're saying that there are cases where an action might be invalid even though the mask says it is valid.

Almost. To be more precise, I say that an action might be invalid even though the mask says it is *not* invalid.

> Is that only for environments with continuous actions spaces or there are other cases?

No, I expect this to happen mostly in environments where computing the mask is expensive, for example the crystal.

To add a bit of an answer to the rest of your comments, which is relevant: I think of all this this way: a vanilla implementation of GFlowNet does not use any masks at all. This is not wrong, it's just slower to train (potentially too slow) because it's missing an important inductive bias and it has to learn that some actions are impossible (ideally masked). One can add _some_ masks of *invalid* actions to make things better. But, it is conceivable to think that we cannot catch all *invalid* masks easily, especially if there are too many actions. That's why the masks is about invalid actions, not about valid actions. One level more is to check, when we attempt to execute one step with one specific action, if the action is indeed valid (it may be cheap enough to check validity for one action, but too expensive to do it for all actions.

All this is mostly theory, I don't mean that all of it is currently implemented, but it's somehow behind some design choices I have made.